### PR TITLE
Botsort tracker

### PR DESCRIPTION
**Pull request by Oleg-1703**

### CODE REVIEW

File: supervisely/nn/model/prediction_session.py [hunk 1]
```diff
@@ -558,7 +558,8 @@ def _predict_videos(
         end_frame=None,
         duration=None,
         direction: Literal["forward", "backward"] = None,
-        tracker: Literal["bot", "deepsort"] = None,
+        tracker: Literal["botsort", "boxmot"] = None,
+        tracker_settings: Dict[str, Any] = None,
```
Comment (max-unfinity) [inline supervisely/nn/model/prediction_session.py]: tracker_settings are not applied, won't work

File: supervisely/nn/model/model_api.py [hunk 2]
```diff
@@ -235,28 +240,29 @@ def predict(
         img_size: int = None,
         classes: List[str] = None,
         upload_mode: str = None,
-        recursive: bool = None,
+        recursive: bool = False,
+        tracking: bool = None,
```
Comment (max-unfinity) [inline supervisely/nn/model/model_api.py:255]: implement activation of tracking algorithm

File: supervisely/nn/tracker/botsort/tracker/tracking_utils/evaluation.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
+import os
+import numpy as np
+import copy
+import motmetrics as mm
+mm.lap.default_solver = 'lap'
+
+from yolox.tracking_utils.io import read_results, unzip_objs
+
+
+class Evaluator(object):
```
Comment (max-unfinity) [inline supervisely/nn/tracker/botsort/tracker/tracking_utils/evaluation.py]: Are this Evaluator used in any of botsort code?

File: supervisely/nn/tracker/botsort/tracker/tracking_utils/io.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
+import os
+from typing import Dict
+import numpy as np
+
+
+def write_results(filename, results_dict: Dict, data_type: str):
```
Comment (max-unfinity) [inline supervisely/nn/tracker/botsort/tracker/tracking_utils/io.py]: are they used in botsort too?

File: supervisely/nn/tracker/botsort/tracker/tracking_utils/timer.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import time
+
+
+class Timer(object):
```
Comment (max-unfinity) [inline supervisely/nn/tracker/botsort/tracker/tracking_utils/timer.py]: are this used in botsort too?

File: supervisely/nn/tracker/botsort/tracker/bot_sort.py
Comment (max-unfinity) [inline supervisely/nn/tracker/botsort/tracker/bot_sort.py:1]: Seems like this file is not used in any of botsort code

File: supervisely/nn/tracker/botsort/__init__.py [hunk 6]
```diff
@@ -0,0 +1,2 @@
+import torch  # pylint: disable=import-error
```
Comment (max-unfinity) [inline supervisely/nn/tracker/botsort/__init__.py]: ???

File: supervisely/nn/tracker/base_tracker.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        """Return the accumulated VideoAnnotation."""
+        raise NotImplementedError("This method should be overridden by subclasses.")
+
+    def _validate_device(self):
+        if self.device != 'cpu' and not self.device.startswith('cuda'):
+            raise ValueError(
+                f"Invalid device '{self.device}'. Supported devices are 'cpu' or 'cuda'."
+            )
+
+    ### methods from another script:
```
Comment (max-unfinity) [inline supervisely/nn/tracker/base_tracker.py]: remove comments

File: supervisely/nn/tracker_legacy/tracker/bot_sort/__init__.py
Comment (max-unfinity) [inline supervisely/nn/tracker_legacy/tracker/bot_sort/__init__.py:1]: Can we remove all tracker_legacy code?

File: supervisely/nn/inference/inference.py [hunk 8]
```diff
@@ -1241,6 +1241,29 @@ def _on_model_deployed(self):
 
     def get_classes(self) -> List[str]:
         return self.classes
+    
+    def _tracker_init(self, state: dict) -> object:
```
Comment (max-unfinity) [inline supervisely/nn/inference/inference.py]: passing "state" is ambiguous, pass only parameters you need, e.g, self, tracker='botsort', tracker_settings=None

File: supervisely/nn/tracker/botsort_tracker.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        matches = []
+        for pair in detection_track_map:
+            det_id = pair["det_id"]
+            track_id = pair["track_id"]
+
+            if track_id is not None:
+                match = {
+                    "track_id": track_id,
+                    "det_id": det_id,
+                    "label": annotation.labels[det_id].to_json()
```
Comment (max-unfinity) [inline supervisely/nn/tracker/botsort_tracker.py]: may be remove to_json()

File: supervisely/nn/tracker/botsort_tracker.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+        with open(config_path, 'r', encoding='utf-8') as file:
+            return yaml.safe_load(file)
+
+    def update(self, frame: np.ndarray, annotation: Annotation) -> List[TrackedObject]:
+        """Update tracker and return tracks for current frame."""
+        self.frame_shape = frame.shape[:2]
+        self._update_obj_classes(annotation)
+        detections = self._convert_annotation(annotation)
+        output_stracks, detection_track_map = self.tracker.update(detections, frame)
+        tracks = self._stracks_to_tracks(output_stracks, detection_track_map)
```
Comment (max-unfinity) [inline supervisely/nn/tracker/botsort_tracker.py:75]: do we realy need this self._stracks_to_tracks?
Ideally we should create video_annotation without using TrackedObject dataclass. This will help getting less bugs in future.

File: supervisely/nn/inference/inference.py [hunk 11]
```diff
@@ -1857,28 +1868,37 @@ def _inference_video(
                 source=frames,
                 settings=inference_settings,
             )
+            
+            if self._tracker is not None:
+                updated_anns = []
+                for frame, ann in zip(frames, anns):
+                    tracks, matches = self._tracker.update(frame, ann)
```
Comment (max-unfinity) [inline supervisely/nn/inference/inference.py]: forgot change this code after updates

File: supervisely/nn/inference/inference.py [hunk 12]
```diff
@@ -2103,34 +2112,47 @@ def _inference_video_id(
...
                 settings=inference_settings,
             )
+            
+            if self._tracker is not None:
+                updated_anns = []
+                for frame, ann in zip(frames, anns):
+                    matches = self._tracker.update(frame, ann)
+ 
+                    track_ids = [match["track_id"] for match in matches]
+                    tracked_labels = [ann.labels[match["det_id"]] for match in matches]
```
Comment (max-unfinity) [inline supervisely/nn/inference/inference.py]: could be done easier?
`tracked_labels = [match["label"] for match in matches]`
In case you do not make `to_json()` there.
Also, det_id is a bit redundant, I think.

File: supervisely/nn/inference/inference.py [hunk 13]
```diff
@@ -1269,7 +1292,7 @@ def get_info(self) -> Dict[str, Any]:
             "async_video_inference_support": True,
             "tracking_on_videos_support": True,
```
Comment (max-unfinity) [inline supervisely/nn/inference/inference.py]: TODO: make it False here; make it True in ObjectDetection and InstanceSegmentation

File: supervisely/nn/inference/session.py [hunk 14]
```diff
@@ -586,7 +586,8 @@ def _on_async_inference_end(self):
     def _post(self, *args, retries=5, **kwargs) -> requests.Response:
         retries = min(self.api.retry_count, retries)
         url = kwargs.get("url") or args[0]
-        method = url[len(self._base_url) :]
+        method = tus
```
Comment (max-unfinity) [inline supervisely/nn/inference/session.py]: ???

File: supervisely/nn/tracker/base_tracker.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+                self.device = auto_device
+            else:
+                self.device = settings_device
+        else:
+            self.device = device if device is not None else auto_device
+                
+        self._validate_device()
+
+
+    def update(self, frame, detections):
```
Comment (max-unfinity) [inline supervisely/nn/tracker/base_tracker.py]: add code annotations: what does this method return? parameter's type. Do it for other methods too

File: supervisely/nn/tracker/botsort_tracker.py [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+        
+        matches = []
+        for pair in detection_track_map:
+            det_id = pair["det_id"]
+            track_id = pair["track_id"]
+
+            if track_id is not None:
+                match = {
+                    "track_id": track_id,
+                    "det_id": det_id,
```
Comment (max-unfinity) [inline supervisely/nn/tracker/botsort_tracker.py]: may be we don't need det_id

File: supervisely/nn/tracker/botsort_tracker.py [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+
+    def _load_default_settings(self) -> dict:
+        """Load default settings from YAML file in the same directory."""
+        current_dir = Path(__file__).parent
+        config_path = current_dir / "botsort/botsort_config.yaml"
+        
+        with open(config_path, 'r', encoding='utf-8') as file:
+            return yaml.safe_load(file)
+
+    def update(self, frame: np.ndarray, annotation: Annotation) -> List[TrackedObject]:
```
Comment (max-unfinity) [inline supervisely/nn/tracker/botsort_tracker.py]: fix code annotation for return type

File: supervisely/nn/model/prediction.py [hunk 18]
```diff
@@ -125,6 +126,16 @@ def _get_confidence(label: Label):
...
         self._boxes = np.array(self._boxes)
         self._masks = np.array(self._masks)
+        
+        self._track_ids = []
+        custom_data = self.annotation.custom_data
+        
+        if custom_data and isinstance(custom_data, list) and len(custom_data) == len(self.annotation.labels):
+            self._track_ids = list(custom_data)
+        else:
+            self._track_ids = [None] * len(self.annotation.labels)
```
Comment (max-unfinity) [inline supervisely/nn/model/prediction.py]: set self._track_ids to None if not relevant

File: supervisely/nn/model/prediction_session.py [hunk 19]
```diff
@@ -111,6 +115,21 @@ def __init__(
         self.inference_settings = {
             k: v for k, v in kwargs.items() if isinstance(v, (str, int, float))
         }
+        
+        if tracking is True:
+            model_info = self._get_session_info()
+            if not model_info.get("tracking_on_videos_support", False) and self.tracking:
```
Comment (max-unfinity) [inline supervisely/nn/model/prediction_session.py]: remvoe ` and self.tracking`

File: supervisely/nn/model/prediction_session.py [hunk 20]
```diff
@@ -67,8 +67,11 @@ def __init__(
         dataset_id: Union[List[int], int] = None,
         project_id: Union[List[int], int] = None,
         api: "Api" = None,
+        tracking: bool = None,
+        tracking_settings: dict = None,
```
Comment (max-unfinity) [inline supervisely/nn/model/prediction_session.py]: rename to tracking_config

File: supervisely/nn/model/prediction_session.py [hunk 21]
```diff
@@ -111,6 +115,21 @@ def __init__(
...
+            model_info = self._get_session_info()
+            if not model_info.get("tracking_on_videos_support", False) and self.tracking:
+                raise ValueError("Tracking is not supported by this model")
+            
+            if tracking_settings is None:
+                self.tracker = "botsort"
+                self.tracker_settings = {}
+            else:
+                self.tracker = tracking_settings.get("tracker", "botsort")
+                self.tracker_settings = tracking_settings.get("settings", {})
```
Comment (max-unfinity) [inline supervisely/nn/model/prediction_session.py]: merge tracker and settings into one dict - tracker_config

File: supervisely/nn/tracker/base_tracker.py [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+                self.device = auto_device
+            else:
+                self.device = settings_device
+        else:
+            self.device = device if device is not None else auto_device
+                
+        self._validate_device()
+
+
+    def update(self, frame, detections) -> List[Dict[str, Any]]:
```
Comment (max-unfinity) [inline supervisely/nn/tracker/base_tracker.py]: add annotations

**Review by max-unfinity — APPROVED** [review]

File: supervisely/nn/model/prediction.py [hunk 23]
```diff
@@ -178,6 +183,12 @@ def class_idxs(self) -> np.ndarray:
             obj_class.name: i for i, obj_class in enumerate(self.model_meta.obj_classes)
         }
         return np.array([cls_name_to_idx[class_name] for class_name in self.classes])
+    @property
+    def track_ids(self):
+        """Get track IDs for each detection. Returns None for detections without tracking."""
+        if self._track_ids is None:
+            self._init_geometries()
```
Comment (max-unfinity) [inline supervisely/nn/model/prediction.py:190]: can retrigger _init_geometries, need flag `self._is_inited`
